File: apps/backend/app/schemas/upload_history.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class UploadHistoryDayResponse(BaseModel):
    date: str = Field(examples=["2026-05-09"])
    upload_count: int = Field(ge=0)
    has_suspected_risk: bool


class UploadHistorySummary28dResponse(BaseModel):
    all_upload_count_28d: int = Field(ge=0)
    suspected_upload_count_28d: int = Field(ge=0)
    continuous_upload_streak_days: int = Field(ge=0, le=28)


class UploadHistoryResponse(BaseModel):
    status: str = Field(examples=["matched", "pending", "unbound"])
    patient_id: int | None
    can_upload: bool
    days: list[UploadHistoryDayResponse]
    summary_28d: UploadHistorySummary28dResponse


class PatientDayUploadItemResponse(BaseModel):
    upload_id: int
    created_at: datetime
    screening_result: Literal["normal", "suspected", "rejected", "technical_error"]
    probability: float | None
    threshold: float | None
    model_version: str | None
    error_reason: str | None
    annotation_label: str | None
    annotation_comment: str | None


class PatientDayUploadListResponse(BaseModel):
    date: str = Field(examples=["2026-05-11"])
    items: list[PatientDayUploadItemResponse]


class PatientUploadDetailResponse(BaseModel):
    upload_id: int
    created_at: datetime
    date: str = Field(examples=["2026-05-11"])
    screening_result: Literal["normal", "suspected", "rejected", "technical_error"]
    probability: float | None
    threshold: float | None
    model_version: str | None
    error_reason: str | None
    annotation_label: str | None
    annotation_comment: str | None
    image_url: str
    image_expires_in: int = Field(ge=1)
    prev_upload_id: int | None
    next_upload_id: int | None

File: apps/frontend/lib/api/upload-history.ts
import { apiClient } from "@/lib/api/client";

export type UploadHistoryDay = {
  date: string;
  upload_count: number;
  has_suspected_risk: boolean;
};

export type UploadHistorySummary28d = {
  all_upload_count_28d: number;
  suspected_upload_count_28d: number;
  continuous_upload_streak_days: number;
};

export type UploadHistoryResponse = {
  status: "matched" | "pending" | "unbound";
  patient_id: number | null;
  can_upload: boolean;
  days: UploadHistoryDay[];
  summary_28d: UploadHistorySummary28d;
};

export type PatientDayUploadItem = {
  upload_id: number;
  created_at: string;
  screening_result: "normal" | "suspected" | "rejected" | "technical_error";
  probability: number | null;
  threshold: number | null;
  model_version: string | null;
  error_reason: string | null;
  annotation_label: string | null;
  annotation_comment: string | null;
};

export type PatientDayUploadListResponse = {
  date: string;
  items: PatientDayUploadItem[];
};

export type PatientUploadDetailResponse = {
  upload_id: number;
  created_at: string;
  date: string;
  screening_result: "normal" | "suspected" | "rejected" | "technical_error";
  probability: number | null;
  threshold: number | null;
  model_version: string | null;
  error_reason: string | null;
  annotation_label: string | null;
  annotation_comment: string | null;
  image_url: string;
  image_expires_in: number;
  prev_upload_id: number | null;
  next_upload_id: number | null;
};

export type PatientMessageItem = {
  annotation_id: number;
  upload_id: number;
  created_at: string;
  label: string;
  comment: string | null;
  is_read: boolean;
  image_url: string;
  image_expires_in: number;
};

export type PatientMessageListResponse = {
  items: PatientMessageItem[];
  total: number;
  unread_count: number;
  limit: number;
  offset: number;
};

export async function fetchUploadHistory(): Promise<UploadHistoryResponse> {
  const { data } = await apiClient.get<UploadHistoryResponse>("/v1/patient/upload-history");
  return data;
}

export async function fetchUploadsByDay(date: string): Promise<PatientDayUploadListResponse> {
  const { data } = await apiClient.get<PatientDayUploadListResponse>("/v1/patient/uploads/by-day", {
    params: { date },
  });
  return data;
}

export async function fetchPatientUploadDetail(uploadId: number): Promise<PatientUploadDetailResponse> {
  const { data } = await apiClient.get<PatientUploadDetailResponse>(`/v1/patient/uploads/${uploadId}/detail`);
  return data;
}

export async function fetchPatientMessages(params?: {
  limit?: number;
  offset?: number;
  unreadOnly?: boolean;
}): Promise<PatientMessageListResponse> {
  const { data } = await apiClient.get<PatientMessageListResponse>("/v1/patient/messages", {
    params: {
      limit: params?.limit ?? 20,
      offset: params?.offset ?? 0,
      unread_only: params?.unreadOnly ?? false,
    },
  });
  return data;
}

export async function markPatientMessageRead(annotationId: number): Promise<PatientMessageItem> {
  const { data } = await apiClient.post<PatientMessageItem>(`/v1/patient/messages/${annotationId}/read`);
  return data;
}
